### ISSUE TITLE
[Feature Request] use rsync to host flatpak repository

### ISSUE THREAD

**Issue by willyywt** [flatpak/flatpak#4383]
I am a user from China and I am affected by https://github.com/flathub/flathub/issues/813, as GFW in China completely censored Fastly CDN.
Currently flatpak repo are only intended to used by a CDN, but it bascially restricted the mirroring method to becoming a reverse proxy, which is proved very toxic to mirror sites: for [tuna](https://github.com/tuna/issues/issues/523) and [sjtug](https://github.com/sjtug/mirror-requests/issues/86)

## Rsync rationale
I don't know flatpak very well, but I think flatpak is a repo manager for an ostree repo, which itself seems to be easy for rsync to mirror: https://github.com/ostreedev/ostree-releng-scripts/blob/master/rsync-repos. 
I suggest flatpak should support an additional remote storage backend using rsync in parallel with the current CDN-like approach.

## Rsync advantages
- Commonly accepted protocol for mirroring 
- Incremental updates
- Packed transfering of files
- Safe resume from failed sync
- Support both unauthorized client and authorized client

## Rsync problems
- Pressure on server resource
- Still more costly than plain deb/rpm packages (but better than making large number of small http requests)

## CDN advantage
- Doesn't require a strong server
- Approachable to many (but not more than mirror sites since CDN providers are not approachable everywhere) areas using a relatively low price

## CDN-only problems
- Effectively limits mirror methods to reverse proxy only, which is unacceptable for some mirror sites and irritating for the rest
- Too many delta file increases the risk of failure because of increased number of http request round trip
- No safe and incremental way of recovering from failed download

These are not flatpak-specific issues; many FOSS projects are affected by them.

Many mirror sites can support rsync in addition to plain http; for these sites rsync are a proved working way of mirroring and are way easier to manage. If remotes like flathub doesn't have much server resource, they can limit the bandwidth to unauthorized clients and reserve more resource to mirror sites.

**Comment by willyywt** [comment]
Flatpak is essentially about software distribution, but it's very different from normal http websites. While using CDN to host websites seems the most straight forward, for FOSS projects to be accessible by people in more areas, we need to take advantage of mirror sites available, which needs a reliable, incremental and commonly used method to sync data.

**Comment by willyywt** [comment]
This is different than https://github.com/flatpak/flatpak/issues/1625 (casync) which is an entirely new thing. Flatpak is very deeply dependent on ostree, but if rsync can pull ostree data easily, then maybe flatpak can add support for rsync, which is very useful when servers (and/or clients) can support rsync.

**Comment by refi64** [comment]
If I'm understanding correctly, this would be allowing the use of `rsync` to synchronize a mirror of Flathub, right? This wouldn't be a Flatpak issue, but I don't believe this is doable regardless for [the reasons listed here](https://github.com/dbnicholson/ostree-push#background) (replace "publish" with "mirror" and the same concepts apply):

> Using rsync for publishing has some major benefits since only updated objects are published. However, it has no concept of the OSTree object store or refs structures. There are a few problems deriving from this issue. First, objects are published in sort order, but this means that objects can be published before their children. In the most extreme case, a commit object could be published before it's complete. The remote repo would assume this commit object was valid even though some children might be missing. Second, the refs might get updated before the commit objects are in place. If a client pulls while rsync is publishing, it may attempt to pull an incomplete or entirely missing commit. Finally, rsync will push the objects directly into the store rather than using a staging directory like pull or commit do. If rsync is interrupted, it could leave partial objects in the store.

In theory, you could have a mirroring tool with knowledge of OSTree internals, but that would be a separate effort.

**Comment by refi64** [comment]
(Aside: that's not to say it wouldn't be possible to have a more efficient mirroring system with knowledge of OSTree, kind of like how `ostree-push` is able to push things efficiently without needing `rsync`, but this would be a separate project and not directly related to Flatpak anyway.)

**Comment by itsnotsaved** [comment]
https://github.com/flatpak/flatpak/issues/4874
